### DISCUSSION TITLE
how to use pre request script properly

### DISCUSSION THREAD

**Discussion by dongokovacs** [usebruno/bruno#992] [Q&A] [answered]
hello guys

I wrote a pre request script:
```
`if (!bru.getEnvVar("ACCESS_TOKEN") || Date.now() >= bru.getEnvVar("ACCESS_TOKEN_EXPIRES"))
{
    req.setMethod({
        url: 'https://login.microsoftonline.com/' + bru.getEnvVar("AZURE_TENANT_ID") + '/oauth2/v2.0/token',
        method: 'POST',
        header: 'Content-Type: multipart/form-data',
        body: {
            mode: 'formdata',
            formdata: [ 
                {key: "grant_type", value: "client_credentials", disabled: false},
                {key: "client_id", value: bru.getEnvVar("AZURE_CLIENT_ID"), disabled: false},
                {key: "client_secret", value: bru.getEnvVar("AZURE_CLIENT_SECRET"), disabled: false},
                {key: "scope", value: bru.getVar("AZURE_TARGET_SCOPE"), disabled: false},
            ]
        }
    }, function (err, res) {
        if (err)
        {
            console.log(err);
        }
        else
        {
            let json = res.json();
            bru.setVar("ACCESS_TOKEN", json.access_token);
            bru.setVar("ACCESS_TOKEN_EXPIRES", Date.now() + json.expires_in * 1000);
        }
    });
}`
```

I use environment variable

but i get an error:
Error invoking remote method 'send-http-request': TypeError: (config.method || this.defaults.method || "get").toLowerCase is not a function

What does this error mean? And have anybody an idea to solve i?

thanks

**Comment by Kaargo1** [comment]
What are you trying to do with the script?
I think your usage of req.setMethod is incorrect as it is only supposed to set what HTTP method your request should use eg. "POST" or "GET".

If you want to set the url, method, body etc you need to use a setter for each of these like req.setUrl, req.setHeader, etc.
Take a look at https://docs.usebruno.com/scripting/javascript-reference.html on how you can use these.

**Comment by dongokovacs** [comment]
I write to this format:
```
const axios = require('axios');

const tokenUrl = 'https://login.microsoftonline.com/' + bru.getEnvVar('AZURE_TENANT_ID') + '/oauth2/v2.0/token';
const clientId = bru.getEnvVar('AZURE_CLIENT_ID');
const clientSecret = bru.getEnvVar('AZURE_CLIENT_SECRET');
const scope = bru.getEnvVar('AZURE_TARGET_SCOPE');


try {
    let resp = await axios({
        method: 'POST',
        url: tokenUrl,
        headers: {
            'Content-Type': 'multipart/form-data',
        },
        data: {
            grant_type: 'client_credentials',
            client_id: clientId,
            client_secret: clientSecret,
            scope: scope
        },
    });

    bru.setVar('ACCESS_TOKEN2', resp.data.access_token);
    bru.setVar('activeToken', clientId + '_' + tokenUrl);
} catch (error) {
    throw error;
}
```

but i get : Error invoking remote method 'send-http-request': AxiosError: Request failed with status code 400


i would like to login to azure ad webapp with bearer token

It is working in postman, but i can not make it to the correct form to bruno

**Reply by dongokovacs** [reply]
my env csope: api://.../.default

**Reply by Kaargo1** [reply]
Status code 400 indicates that there is something wrong with the request you are sending.
Double-check the url, headers and request body and make sure that its the same as in Postman

**Reply by dongokovacs** [reply]
It will work with simple get
v1: simple - set auth to bearer, and copy token from webapp console network tab OK

v2: pre csript - in my pre script request i set the auth to bearer token: {{ACCESS_TOKEN}}
my scope: api://..../.default

in postman i can login with this credentials

**Reply by Kaargo1** [reply]
I'm not sure I follow anymore. Are you still having issues?
Could you perhaps provide some code and/or screenshots of what the issue is?

**Reply by dongokovacs** [reply, marked as answer]
i am not sure that u could help me more
my response: 401
response bodi: empty
auth: bearer: {{ACCESS_TOKEN}}

my code:
```
const axios = require('axios');

const tokenUrl = 'https://login.microsoftonline.com/' + bru.getEnvVar('AZURE_TENANT_ID') + '/oauth2/v2.0/token';
const clientId = bru.getEnvVar('AZURE_CLIENT_ID');
const clientSecret = bru.getEnvVar('AZURE_CLIENT_SECRET');
const scope = bru.getEnvVar('AZURE_TARGET_SCOPE');

console.log(clientId)
console.log(clientSecret)
console.log(scope)

try {
    let resp = await axios({
        method: 'POST',
        url: tokenUrl,
        headers: {
            'Content-Type': 'multipart/form-data',
        },
        data: {
            grant_type: 'client_credentials',
            client_id: clientId,
            client_secret: clientSecret,
            scope: scope
        },
    });
  console.log(resp)
    bru.setVar('ACCESS_TOKEN2', resp.data.access_token);
    bru.setVar('activeToken', clientId + '_' + tokenUrl);
} catch (error) {
    throw error;
}`
````const axios = require('axios');

const tokenUrl = 'https://login.microsoftonline.com/' + bru.getEnvVar('AZURE_TENANT_ID') + '/oauth2/v2.0/token';
const clientId = bru.getEnvVar('AZURE_CLIENT_ID');
const clientSecret = bru.getEnvVar('AZURE_CLIENT_SECRET');
const scope = bru.getEnvVar('AZURE_TARGET_SCOPE');

console.log(clientId)
console.log(clientSecret)
console.log(scope)

try {
    let resp = await axios({
        method: 'POST',
        url: tokenUrl,
        headers: {
            'Content-Type': 'multipart/form-data',
        },
        data: {
            grant_type: 'client_credentials',
            client_id: clientId,
            client_secret: clientSecret,
            scope: scope
        },
    });
  console.log(resp)
    bru.setVar('ACCESS_TOKEN', resp.data.access_token);
    bru.setVar('activeToken', clientId + '_' + tokenUrl);
} catch (error) {
    throw error;
}

```

**Reply by dongokovacs** [reply]
Can i sendRequest without axios?

**Reply by Kaargo1** [reply]
Bruno has support for installing external libraries via npm so you might be able to install something from there. According to the docs you can also use node-fetch which seems to be a built-in library. However, I don't have any experience using anything other than Axios in Bruno.

Side-note:
I took a look at the Microsoft docs for getting the [access token for Azure AD via API](https://learn.microsoft.com/en-us/azure/databricks/dev-tools/service-prin-aad-token#--get-a-microsoft-entra-id-access-token-with-the-microsoft-identity-platform-rest-api) and it's possible that your Content-type header should be changed to: application/x-www-form-urlencoded

Unfortunately I don't have access to Azure and can't try anything out to see if I have the same problem but if you have any more questions I'll do my best to help.

**Reply by dongokovacs** [reply]
I did it finally:
const axios = require('axios');

const tokenUrl = 'https://login.microsoftonline.com/' + bru.getEnvVar('AZURE_TENANT_ID') + '/oauth2/v2.0/token';
const clientId = bru.getEnvVar('AZURE_CLIENT_ID');
const clientSecret = bru.getEnvVar('AZURE_CLIENT_SECRET');
const scope = bru.getEnvVar('AZURE_TARGET_SCOPE');

const axiosInstance = axios.create({
  baseURL: tokenUrl,
  headers: {
    'Content-Type': 'application/x-www-form-urlencoded',
  },
});

const data = {
  grant_type: 'client_credentials',
  client_id: clientId,
  client_secret: clientSecret,
  scope: scope,
};

axiosInstance.post('', data)
  .then(response => {
    const accessToken = response.data.access_token;
    console.log('Authentication successful!');
    console.log('Access token:', accessToken);

    // Use the access token for subsequent API requests
  bru.setEnvVar('ACCESS_TOKEN', accessToken);
  })
  .catch(error => {
    console.error('Authentication failed:', error);
  });

**Reply by Kaargo1** [reply]
Nice! Glad it worked out.
Do you know if it worked because you changed the Content-Type header or because you changed the way you use Axios?

**Reply by ericdain** [reply]
Yes, I wish @daniavander would have followed up with exactly how they got it to work. This thread has been helpful to me, but I'm still not all the way there. It would help many others to have this scenario (pre-request script to fetch a bearer token for use in subsequent API calls) well documented!

I've tested my pre-request script outside of Bruno so I know the Bearer token request is working as expected, and I've manually confirmed the generated token works in Bruno if I place it directly into the Auth > (Bearer) Token field. But I get a "malformed access token: Not enough or too many segments" error when I try to use "{{ACCESS_TOKEN}}" in the Auth > Token field.

~~It would be helpful if there was a way to get "console.log" output in Bruno, to confirm things are what I expect from within Bruno, but I haven't figured that out yet.~~ Found the developer tools under the View menu!

**Comment by ericdain** [comment]
I finally got this working. A few things that tripped me up that might help others:

- You can see Console.log output by opening the "Developer Tools" under the "View" file menu. Very helpful!
- `bru.setEnvVar` is different than `bru.setVar`. I started off using `setEnvVar`, thinking this would access and set variables within the "Bruno" environment, like those under the "Vars" tab. But this doesn't seem to be the case--use `setVar` for that. [The docs](https://docs.usebruno.com/scripting/javascript-reference#environments) touch on this but are not as clear/verbose as I would like.
- With my pre-request script and assigning a response value to variable `ACCESS_TOKEN` for subsequent use, I find that while the value does get saved (you can tell by the color of the token value placeholder `{{ACCESS_TOKEN}}` changing from a red color to green, and hovering also shows the current value in a tooltip), the subsequent API request fails the FIRST TIME. Afterwards however, now that the ACCESS_TOKEN variable is saved (and it persists), my "main" API request now works as expected by using the previously set token variable. I feel like it should work the first time, but alas, it seems like there is a race condition. I found that adding a sleep of 0.5 seconds to the end of the pre-request script did cause the FIRST "main" request to work.
- Once a variable is set, I found it hard to change it. I started completely restarting Bruno to clear everything when needed to help ensure nothing was being persisted that I wasn't expecting.

```javascript
const axios = require('axios');

const tokenUrl = bru.getVar('TOKEN_URL');
const clientId = bru.getVar('CLIENT_ID');
const clientSecret = bru.getVar('CLIENT_SECRET');
const scope = bru.getVar('CLIENT_SCOPE');

const axiosInstance = axios.create({
  baseURL: tokenUrl,
  headers: {
  'Content-Type': 'application/x-www-form-urlencoded',
  },
});

const data = {
grant_type: 'client_credentials',
client_id: clientId,
client_secret: clientSecret,
scope: scope,
};

axiosInstance.post('', data).then(response => {
  const accessToken = response.data.access_token;
  console.log('Authentication successful!');
  console.log('Access token:', accessToken);

  // Use the access token for subsequent API requests
  bru.setVar('ACCESS_TOKEN', accessToken);
})
.catch(error => {
console.error('Authentication failed:', error);
});
```

Hope this helps someone!

**Reply by sguerrero11** [reply]
hey, had a qq here: if you set an Env Var in the pre request script, can you access that variable in the request itself? it seems it's not being persisted. thoughts?

**Reply by sterankin** [reply]
This is a great help.  Some additions - it might be nice to check there is a token var already and check the expiration time and only get a new token if its expired.

**Reply by ericdain** [reply]
> hey, had a qq here: if you set an Env Var in the pre request script, can you access that variable in the request itself? it seems it's not being persisted. thoughts?

yes you should be able to access a set env var, otherwise it seems kind of pointless. In my example I pass it as the value for the Authorization header along with the request. I definitely had to play with it a lot to uncover things I noted above. I haven't used this much since posting, so I'm not familiar with any changes that might have been implemented since then, though.

> This is a great help. Some additions - it might be nice to check there is a token var already and check the expiration time and only get a new token if its expired.

Yep so true!

**Reply by sterankin** [reply]
This works you just need to do a check in the script of the token is already in the env var or so, if it is there check the expiry:

```
function isJwtExpired(token) {
  if (typeof(token) !== 'string' || !token || !token.includes(".")) {
    throw new Error('Invalid token provided');
  }

  const payload = Buffer.from(token.split('.')[1], 'base64').toString('ascii');
  const { exp } = JSON.parse(payload);

  const currentTime = new Date().getTime() / 1000;

  if (currentTime > exp) {
    console.log('JWT expired!');
    return true;
  }
  return false;
}
```

**Reply by dineshbhagat** [reply]
Using inbuilt libs [Ref](https://docs.usebruno.com/testing/script/inbuilt-libraries)  
Same as above method and replace `const payload = Buffer.from(token.split('.')[1], 'base64').toString('ascii');` with
```javascript
const atob = require('atob');
const payload = atob(token.split('.')[1])
```

**Comment by ayumakesit** [comment]
Hi guys,
@ericdain guided me to the right point, thanks!

I had similar problems:

1. I added a "Pre Request Var" to the collection
2. The pre request script ran successfully, but didn't save the value to the variable correctly.
3. After adding `const sleep = ms => new Promise(r => setTimeout(r, ms)); await sleep(2000);` to the pre request script, the value was saved correctly to the variable
4. After that I changed the environment to my staging domain and now the request fails again

Looks like a bug, right?

Thanks,
